### PR TITLE
architectures: add schema support

### PR DESCRIPTION
**Pull request by mr-cal**
- [X] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [X] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [X] Have you successfully run `make lint`?
- [X] Have you successfully run `pytest tests/unit`?

-----
Add support for defining architectures in snapcraft.yaml.

Part 1 of supporting architectures for core22.  Full code change is proposed [here](https://github.com/snapcore/snapcraft/pull/3712).

(CRAFT-1163)

### CODE REVIEW

File: snapcraft/projects.py [hunk 1]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
 
 
+def _validate_architectures(architectures):
+    """Expand and validate architecture data.
+
+    Validation includes:
+        - The list cannot be a combination of strings and Architecture objects.
+        - The same architecture cannot be defined in multiple `build-to` fields,
+          even if the implicit values are used to define `build-to`.
+        - Only one architecture can be defined in the `build-to` list.
```
Comment (sergiusens) [inline snapcraft/projects.py:75]:
```suggestion
        - Only one architecture can be defined in the `build-for` list.
```

File: snapcraft/projects.py [hunk 2]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+        all(isinstance(architecture, str) for architecture in architectures)
+        or all(isinstance(architecture, Architecture) for architecture in architectures)
+    ):
+        raise ValueError(
+            f"Every item must either be a string or an object for {architectures!r}"
+        )
+
+    _expand_architectures(architectures)
+
+    # validate `build-to` after expanding data
```
Comment (sergiusens) [inline snapcraft/projects.py:91]:
```suggestion
    # validate `build-for` after expanding data
```

File: snapcraft/projects.py [hunk 3]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
     return command_chains
 
 
+def _validate_architectures(architectures):
+    """Expand and validate architecture data.
+
+    Validation includes:
+        - The list cannot be a combination of strings and Architecture objects.
+        - The same architecture cannot be defined in multiple `build-to` fields,
```
Comment (sergiusens) [inline snapcraft/projects.py:73]:
```suggestion
        - The same architecture cannot be defined in multiple `build-for` fields,
```

File: snapcraft/projects.py [hunk 4]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+        or all(isinstance(architecture, Architecture) for architecture in architectures)
+    ):
+        raise ValueError(
+            f"Every item must either be a string or an object for {architectures!r}"
+        )
+
+    _expand_architectures(architectures)
+
+    # validate `build-to` after expanding data
+    if any(len(architecture.build_to) > 1 for architecture in architectures):
```
Comment (sergiusens) [inline snapcraft/projects.py:92]:
```suggestion
    if any(len(architecture.build_for) > 1 for architecture in architectures):
```

**Comment by mr-cal** [comment]
Renamed `build_to` -> `build_for`

File: snapcraft/projects.py [hunk 5]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+
+    :raise ValueError: If architecture data is invalid.
+    """
+    # validate strings and Architecture objects are not mixed
+    if not (
+        all(isinstance(architecture, str) for architecture in architectures)
+        or all(isinstance(architecture, Architecture) for architecture in architectures)
+    ):
+        raise ValueError(
+            f"Every item must either be a string or an object for {architectures!r}"
```
Comment (sergiusens) [inline snapcraft/projects.py:86]: this one is in upper case

File: snapcraft/projects.py [hunk 6]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+        )
+
+    _expand_architectures(architectures)
+
+    # validate `build_for` after expanding data
+    if any(len(architecture.build_for) > 1 for architecture in architectures):
+        raise ValueError("multiple architectures are defined for one 'build_for'")
```
Comment (sergiusens) [inline snapcraft/projects.py:93]:
```suggestion
        raise ValueError("only one architecture can be defined for 'build_for'")
```
```suggestion
        raise ValueError("multiple architectures are defined for one 'build_for'")
```

File: snapcraft/projects.py [hunk 7]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+
+    return architectures
+
+
+def _expand_architectures(architectures):
+    """Expand architecture data.
+
+    Expansion to fully-defined Architecture objects includes the following:
+        - strings (shortform notation) are converted to Architecture objects
+        - `build-on` and `build_for` strings are converted to single item lists
```
Comment (sergiusens) [inline snapcraft/projects.py:116]:
```suggestion
        - `build-on` and `build-for` strings are converted to single item lists
```

File: snapcraft/projects.py [hunk 8]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+        - Empty `build_for` fields are implicitly set to the same architecture used in `build-on`
```
Comment (sergiusens) [inline snapcraft/projects.py:117]:
```suggestion
        - Empty `build-for` fields are implicitly set to the same architecture used in `build-on`
```

File: snapcraft/projects.py [hunk 9]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+            architectures[index] = Architecture(
+                build_on=[architecture], build_for=[architecture]
+            )
+        elif isinstance(architecture, Architecture):
+            # convert strings to lists
+            if isinstance(architecture.build_on, str):
+                architectures[index].build_on = [architecture.build_on]
+            if isinstance(architecture.build_for, str):
+                architectures[index].build_for = [architecture.build_for]
+            # implicitly set build_for to build_on
```
Comment (sergiusens) [inline snapcraft/projects.py:131]:
```suggestion
            # implicitly set build_for from build_on
```

File: snapcraft/projects.py [hunk 10]
```diff
@@ -64,6 +65,99 @@ def _validate_command_chain(command_chains: Optional[List[str]]) -> Optional[Lis
...
+                architectures[index].build_for = architectures[index].build_on
+
+
+def _validate_architectures_all_keyword(architectures):
+    """Validate `all` keyword is properly used.
+
+    Validation rules:
+    - `all` cannot be used to `build-on`
+    - If `all` is used for `build_for`, no other architectures can be defined
+      for `build_for`.
```
Comment (sergiusens) [inline snapcraft/projects.py:142]:
```suggestion
    - If `all` is used for `build-for`, no other architectures can be defined
      for `build-for`.
```

**Review by sergiusens — APPROVED** [review]
added some minor comments

**Pull request merged**